src/unity/toolkits/tcmps/mps_weight.mm

```diff
@@ -9,7 +9,116 @@
 
 using turi::mps::OptimizerOptions;
 
+NS_ASSUME_NONNULL_BEGIN
 
+// Protocol generalizing across MPSNNOptimizerStochasticGradientDescent and
+// MPSNNOptimizerAdam, for convenience.
+API_AVAILABLE(macos(10.14))
+@protocol TCMPSConvolutionWeightsOptimizing <NSObject>
+
+// Conv update
+-(void)encodeToCommandBuffer:(id<MTLCommandBuffer>)commandBuffer
+    convolutionGradientState:(MPSCNNConvolutionGradientState *)gradientState
+          convolutionWeights:(TCMPSConvolutionWeights *)weights;
+
+// Batch norm update
+-(void)encodeToCommandBuffer:(id<MTLCommandBuffer>)commandBuffer
+     batchNormalizationState:(MPSCNNBatchNormalizationState *)state
+               batchNormData:(TCMPSBatchNormData *)data;
+
+-(void)setLearningRate:(float)newLearningRate;
+
+@end
+
+// Category adding TCMPSConvolutionWeightsOptimizing to MPSNNOptimizerAdam
+@interface MPSNNOptimizerAdam (TCMPSWeightsOptimizing) <TCMPSConvolutionWeightsOptimizing>
+@end
+
+// Category adding TCMPSConvolutionWeightsOptimizing to
+// MPSNNOptimizerStochasticGradientDescent
+@interface MPSNNOptimizerStochasticGradientDescent (TCMPSWeightsOptimizing) <TCMPSConvolutionWeightsOptimizing>
+@end
+
+// Class extension for TCMPSConvolutionWeights internal implementation
+@interface TCMPSConvolutionWeights ()
+
+@property (nonatomic, readonly) id<TCMPSConvolutionWeightsOptimizing> optimizer;
+
+@property (nonatomic, readonly) MPSVector *weightMomentumVector;
+@property (nonatomic, readonly) MPSVector *weightVelocityVector;
+@property (nonatomic, readonly) MPSVector *biasMomentumVector;
+@property (nonatomic, readonly) MPSVector *biasVelocityVector;
+
+@end
+
+// Class extension for TCMPSBatchNormData internal implementation
+@interface TCMPSBatchNormData ()
```

Comment:
Should we perhaps call this `TCMPSBatchNormWeights`? Not sure why we have convolution *weights* but batch norm *data*. I suppose the BN gammas and betas are what we normally consider to be weights, but the moving mean/var don't typically go by this name. However, at the end of the day, during inference, they are weights in the network, so the distinction is too subtle to warrant the -Data name.

Reply:
Good call! I went ahead and renamed to `TCMPSBatchNormWeights`